### PR TITLE
A Numba dof transformations function

### LINKED ISSUE #97: Expose cell::str_to_type and cell::type_to_str to python layer
To make it as easy as possible to use basix along-side dolfin-x, I suggest that `cell:str_to_type` and `cell::type_to_str` is expose, such that one can easily convert between `dolfinx::mesh::CellType` and `basix::CellType`.
This is for instance relevant when creating custom quadrature rules, as `make_quadrature` require a `basix` cell type. 
In dolfin-x, one can convert the `dolfinx::mesh::CellType` to a string using `dolfinx.cpp.mesh.to_string(mesh.topology.cell_type)`

### PR DESCRIPTION
**Pull request by mscroggs**
This helper function will allow people writing custom assemblers in DOLFINx to use dof transformations without having to implement from scratch.

Fixes #97.

### CODE REVIEW

File: python/basix/numba_helpers.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+try:
+    import numba
+except ImportError:
+    raise RuntimeError("You must have numba installed to use the numba helper functions.")
+
+
+@numba.njit
+def apply_dof_transformation(entity_transformations, cell_type, entity_dofs,
```
Comment (garth-wells) [inline python/basix/numba_helpers.py:8]: Could you add a docstring? It's not so clear which arguments are in and which are out.

File: python/basix/numba_helpers.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                             data, block_size, cell_info):
+    if cell_type == "interval":
```
Comment (garth-wells) [inline python/basix/numba_helpers.py:10]: String comparison isn't so elegant - maybe return different functions depending on the cell type which can then be used without the internal check.

File: python/basix/numba_helpers.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            face_start = 3 * face_count
+        else:
+            face_start = 0
+
+        dofstart = 0
+        for i in entity_dofs[0]:
+            dofstart += i
+
+        for e in range(edge_count):
+            edofs = entity_dofs[1][e]
```
Comment (garth-wells) [inline python/basix/numba_helpers.py:38]: What type does `entity_dofs` normally have? (Will it be fast?)

**Review by garth-wells — COMMENTED** [review]
Nice, just a few inline comments on docs and performance.

**Comment by michalhabera** [comment]
But why is not the c++ `apply_dof_transformation` exposed and used in python?

**Comment by mscroggs** [comment]
> But why is not the c++ `apply_dof_transformation` exposed and used in python?

I don't think you can call that from inside a numba compiled kernel?

**Comment by michalhabera** [comment]
> > But why is not the c++ `apply_dof_transformation` exposed and used in python?
> 
> I don't think you can call that from inside a numba compiled kernel?

Right. Its just very dangerous to have the same algorithm implemented in two places, error prone. This would be possible if basix had C interface for methods which are meant to be called in the tight loops in custom assemblers.

**Comment by jorgensd** [comment]
> > > But why is not the c++ `apply_dof_transformation` exposed and used in python?
> > 
> > 
> > I don't think you can call that from inside a numba compiled kernel?
> 
> Right. Its just very dangerous to have the same algorithm implemented in two places, error prone. This would be possible if basix had C interface for methods which are meant to be called in the tight loops in custom assemblers.

We had a C interface until recently. The problem is that it was not tested, that causes new issues when implementing new functionality.
The dof permutation code is fairly short, and should not be too hard to maintain. I can add a test to dolfin-x using this functionality to assemble mass matrices if we want to test its correctness.

**Comment by michalhabera** [comment]
> > > > But why is not the c++ `apply_dof_transformation` exposed and used in python?
> > > 
> > > 
> > > I don't think you can call that from inside a numba compiled kernel?
> > 
> > 
> > Right. Its just very dangerous to have the same algorithm implemented in two places, error prone. This would be possible if basix had C interface for methods which are meant to be called in the tight loops in custom assemblers.
> 
> We had a C interface until recently. The problem is that it was not tested, that caused new issues when implementing new functionality in dolfin-x.
> The dof permutation code is fairly short (~50 lines of code), and should not be too hard to maintain. I can add a test to dolfin-x using this functionality to assemble mass matrices if we want to test its correctness.

It is tested against standard c++ code, so that is safe. For the actual permutation case this is not an issue, but it shows an issue in general. Consider as a next step I'd like to call basix basis evaluation inside my custom numba assembler - which is very much desired and one of the advantages for runtime tabulation library, right? Will you reimplement basis evaluation in numba?

If there is a way to call basix methods from numba directly this would be neat.

File: python/basix/numba_helpers.py (hunk 3 above)
Comment (jorgensd) [reply python/basix/numba_helpers.py:43]: entity_dofs[i] is an array of ints. Input type into function has been changed in the doc strings

**Review by jorgensd — APPROVED** [review]

**Comment by mscroggs** [comment]
I've made an issue for wider Numba support: #201

**Pull request merged**